components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,192 @@
+      /*.always_use_initial_delay =*/true);
+  post_results_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(post_results_policy_.get());
+
+  init_task_timer_ = std::make_unique<base::OneShotTimer>();
+  init_task_timer_->Start(
+      FROM_HERE, base::Seconds(init_federated_service_wait_time_in_seconds),
+      this, &LearningService::Init);
+}
+
+LearningService::~LearningService() {
+  StopParticipating();
+  eligibility_service_->RemoveObserver(this);
+}
+
+void LearningService::Init() {
+  DCHECK(url_loader_factory_);
+  DCHECK(eligibility_service_);
+  DCHECK(init_task_timer_);
+  communication_adapter_ =
+      std::make_unique<CommunicationAdapter>(url_loader_factory_);
+  eligibility_service_->AddObserver(this);
+
+  StartParticipating();
+}
+
+void LearningService::StartParticipating() {
+  participating_ = true;
+
+  GetTasks();
+}
+
+void LearningService::StopParticipating() {
+  DCHECK(participating_);
+  participating_ = false;
+}
+
+void LearningService::GetTasks() {
+  communication_adapter_->GetTasks(base::BindOnce(
+      &LearningService::HandleTasksOrReconnect, base::Unretained(this)));
+}
+
+void LearningService::HandleTasksOrReconnect(TaskList tasks, int reconnect) {
+  if (tasks.size() == 0) {
+    reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    reconnect_timer_->Start(FROM_HERE, base::Seconds(reconnect), this,
+                            &LearningService::GetTasks);
+    VLOG(2) << "No tasks available, reconnecting in " << reconnect << "s";
+    return;
+  }
+
+  // TODO(lminto): for now, disregard all tasks beyond the first one
+  // TODO(stevelaskaridis): extract into a utility function
+  Task task = tasks.at(0);
+  float lr = 0.01;
+  std::map<std::string, float> config = task.GetConfig();
+  auto cursor = config.find("lr");
+  if (cursor != config.end()) {
+    lr = cursor->second;
+    VLOG(2) << "Learning rate applied from server: " << lr;
+  }
+
+  ModelSpec spec{
+      32,   // num_params
+      64,   // batch_size
+      lr,   // learning_rate
+      500,  // num_iterations
+      0.5   // threshold
+  };
+
+  if (static_cast<int>(task.GetParameters().at(0).size()) != spec.num_params &&
+      static_cast<int>(task.GetParameters().at(1).size()) != 1) {
+    VLOG(2) << "Task specifies a different model size than the client";
+    return;
+  }
+  VLOG(2) << "Task model and client model match!";
+
+  std::unique_ptr<Model> model = std::make_unique<Model>(spec);
+  model->SetWeights(task.GetParameters().at(0));
+  model->SetBias(task.GetParameters().at(1).at(0));
+  auto task_runner =
+      std::make_unique<FederatedTaskRunner>(task, std::move(model));
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      {base::TaskPriority::BEST_EFFORT,
+       base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN},
+      base::BindOnce(&LearningService::LoadDatasetAndRunTask,
+                     base::Unretained(this), std::move(task_runner)),
```

Comment:
`Unretained` is definitely can't be used here (results it crashes & data races).
Please make `LoadDatasetAndRunTask` a free function (not a class member)

Reply:
I'm seeing the following error when I try to use weakptr: `weak_ptrs can only bind to methods without return values`